ConsoleApplication2 - Test.cpp - discussion

bombela opened the pull request.

buchi68-0 committed:
version with a small modification

conv function : removes 48 from each characters in final_variable so it doesn't have to add 24 back each time we divide by 2 (instead of having O(n)(n+1)/2 complexity (for this part only) it's having O(n) now (n being the length of final_variable ; constant c = 1 (because we're talking about one addition/substraction))

buchi68-0 committed:
version with multiplication

I has the changes from // - Test.cpp ; the multiplication is using the naive method, but I still think it's kinda fast

buchi68-0 committed:
v1.0 - can do multiplication & additions

What it could do :
1 - convert a base10 representation of a number in a char* (i.e. : "234") to a Number ; Numbers can go up to infinity (until not enough RAM), but they can't be (for now) negative
     (Complexity : O(n²) (repetitive divisions by 2) 
2 - Make additions between 2 Numbers. Complexity : O(n)
3 - Convert a Number to a base16 representation. Complexity : O(n)
4 - Make multiplications between 2 Numbers. Complexity : O(n²+48n) (I hope to optimize it a little before implementing the Karatsuba algorithm)

Added to this version :
1 - Added a way to convert Number back to readable base10 char*
    Complexity : O(n²) (repetitive divisions by 10)

Will add in the future :
1 - Add substraction, as well as negative Numbers (signed Number will have a 'flag' (0 or 1) that determines if negative or not)
2 - Maybe add divisions between Numbers (the algorithm for Numbers < 256 is east, but way harder for Numbers that takes more than one byte)

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+#endif
+
+typedef struct Number {
+    uint8_t* Value;
```
Value -> limbs

The use of "limb" in this context can be traced back to at least May 1996 (https://gmplib.org/manual/Nomenclature-and-Types):

> A limb means the part of a multi-precision number that fits in a single machine word. (We chose this word because a limb of the human body is analogous to a digit, only larger, and containing several digits.) Normally a limb is 32 or 64 bits. The C data type for a limb is mp_limb_t.

buchi68-0 replied:
the problem I had when choosing a 64 bits limb, is that there's no 128 bits variable to store the multiplication of the 2 numbers (or maybe just i don't know how to do so) ; that's why i sticked to a 32bit limb on v1.1
Sorry for not responding before

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+
+typedef struct Number {
+    uint8_t* Value;
+    uint32_t length;
```
btw, if you find `length` annoying to type, `len` is a mighty fine alternative.

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+
+Number Init_Number(int base_length) {
+    Number returned = { (uint8_t*)calloc(base_length,1), base_length };
+    if (returned.Value == NULL) {
```
Since 2023, C has gotten `nullptr` (https://en.cppreference.com/w/c/language/nullptr). If your toolchain doesn't support it ignore this.

`if (r.limbs == nullptr) {...`

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+    uint32_t length;
+};
+
+Number Init_Number(int base_length) {
```
`Number new_number(uint32_t len)`

Why not the word init? Because you are creating a fresh number and returning it, not initializing something passed via parameter.

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+};
+
+Number Init_Number(int base_length) {
+    Number returned = { (uint8_t*)calloc(base_length,1), base_length };
```
A convention for returned value is to name is `r`. Its like `i` for indexing in loops etc, some just short conventional names.

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+Number Init_Number(int base_length) {
+    Number returned = { (uint8_t*)calloc(base_length,1), base_length };
+    if (returned.Value == NULL) {
+        exit(90);
```
An abort is more appropriate. https://en.cppreference.com/w/c/program/abort

This is because `abort` clearly signals a failure of the program.

Good job on testing for failed allocation.

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+    return returned;
+}
+
+uint64_t uclen(unsigned char* input_) {
```
`strlen` is the conventional name. https://en.cppreference.com/w/c/string/byte/strlen

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+int main()
+{
+    if (CConsole) CreateConsoleWindow();
+    char a[16001] = "121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891121732378927932891 ...
```
`char* a = ` would work fine here. 

The string literal "4242" is allocated by the compiler into the readonly area of your program. Then you ask it to copy it into an array allocated on the stack. Now, they are cases where its useful. An array knows its size at compile time: `sizeof a -> 16001`.

But you probably do not want/need to allocate such an amount of data on your stack. Nor do you need to copy it either. `char*` will give you a pointer to this read only memory instead. And you can then call `strlen` at runtime to compute the length. And as a bonus, you don't need to manually keep track of the size of your string manually.

For info, most programing languages represent strings as a tuple of (data*, len). Because keeping the length along is simpler and the extra memory consumed with the length variable itself is small comparatively to how big our modern computers are.

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+    }
+
+    //calcul du temps que ça à pris (simple soustraction)
+    long final_time = new_time.tv_sec-my_time.tv_sec ;
```
Its usually nicer to read when adding spaces around the math operators. And avoiding random spaces. Because some languages like C are tricky, often a small visual disturbance indicates something potentially tricky. So we try to maintain some consistency in hope of saving our souls :D

`long final_time = new_time.tv_sec - my_time.tv_sec;`

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+
+//Partie 1 : faire la multiplication simple
+//Il faudra lire les bits directement (sinon la v rification du nombre de 1 sera compliqu e car elle devra se faire   la m thode "classique")
+Number conv(char* _Input) {
```
`parse_decimal` `parse_decimal_number`

`char* str` or `char* input_str`.

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+//Partie 1 : faire la multiplication simple
+//Il faudra lire les bits directement (sinon la v rification du nombre de 1 sera compliqu e car elle devra se faire   la m thode "classique")
+Number conv(char* _Input) {
+    unsigned char* final_variable = (unsigned char*)_Input; //convertion de _Input (variable d'entrée) en un truc plus utilisable
```
What you are doing is manipulating the input string as a string of bytes.
`uint8_t *input = (uint_t*) str`;

C doesn't allow you to re-declare a variable. Otherwise I would want to do `uint8_t *str = (uint_t*) str;`.

Btw this cast only one thing: tell the compiler that you are using bytes, therefor × and ÷ will be unsigned. (addition/subtraction is the same between signed/unsigned because of two's complement).

> signed char — type for signed character representation. 
    unsigned char — type for unsigned character representation. Also used to inspect [object representations](https://en.cppreference.com/w/cpp/language/object) (raw memory). 
    char — type for character representation which can be most efficiently processed on the target system (has the same representation and alignment as either signed char or unsigned char, but is always a distinct type). [Multibyte characters strings](https://en.cppreference.com/w/cpp/string/multibyte) use this type to represent code units. For every value of type unsigned char in range [​0​, 255], converting the value to char and then back to unsigned char produces the original value.(since C++11) The signedness of char depends on the compiler and the target platform: the defaults for ARM and PowerPC are typically unsigned, the defaults for x86 and x64 are typically signed.
    
 source: https://en.cppreference.com/w/cpp/language/types

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+    
+    // verification de final_variable (si il y a autre chose que des chiffres)
+
+    int a = uclen(final_variable); //évite de devoir faire des appels succesifs de uclen
```
By the way, notice that `strlen` returns an `int`, which is different than `uint32_t`. On your computer, `int` is a signed 32 bits (some CPUs are 8, 16, 32 or even 64 bits ints). In your case you can count to something around 2GB of RAM's worth of string.

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+    for (int i = 0; i < a; i++) { 
+        if ((48 > final_variable[i] || final_variable[i] > 57)) {
+            Number returned = { NULL, 0 };
+            return returned; //ce n'est pas un chiffre on retourne un Number vide
```
Alternatively, you could print an error msg and then call `abort`. Instead of pretending the input was `0`. Could be a source of annoying bugs!

A proper implementation would return a Number **or** an error. So that you can tell if the parsing was successful. That's also why the name `parsing` is appropriate here. Because you are attempting to parse the input, which cannot be guaranteed to be correct after all.

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+            return returned; //ce n'est pas un chiffre on retourne un Number vide
+        }
+
+        final_variable[i] -= 48; //on sait que c'est un chiffre, on va donc enlever 48 au code. Même si c'est un zero, c'est pas grave : on a déjà a pour longeure (réele) de final_variable
```
Btw,  you can write `final_variable[i] -= '0'`. This is exactly the same code, but for the reader, this confirms we are interested in ASCII digits.

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+        //cette manipulation permet de ne pas rajouter 24 à final_variable[i] quand on a divisé par 2
+
+        if (count_z && final_variable[i] == 0) {
+            number_of_zeros += 1; // ici on va pré-compter le nombre de 0 depuis la gauche (car si quelqu'un mets "001234", ça va pas marcher)
```
why would this not work? "001234" I mean the leading zeros are useless, but also shouldn't change a thing.

bombela commented on ConsoleApplication2 - Test.cpp:
```diff
@@ -0,0 +1,253 @@
+        if (count_z && final_variable[i] != 0) count_z = false;
+    }
+    
+    uint64_t length = strlen(_Input)-number_of_zeros; //longeur de returned_int (calculée en avance) (on fait attention au number_of_zero) pour ne pas utiliser trop d'espace
```
If I get that right, here you are calculating number of bytes required to store the base10 input as a base2 number.

The math formula is something like `binary digits count = ceil(log10(decimal digits count) / log10(2))`.
We already know the number of digits, so that gets us to `ceil(decimal_digits_count / log10(2))` . Converted to a multiplication, `ceil(digits10 × (1/log10(2)))`. And `1/log10(2) ~= 3.32192809488736234786`.

So if we take digits10 × 4 we are guaranteed to have enough bits, but with some serious waste. What about `digits10 × 34 / 10`. Nice, more precision, less waste. `digits10 × 333 / 100`. Even better. We are simply moving the decimal point to gain more resolution. The limit is how high of a number you can compute. Which in turn limits the maximum number of digits10 you can accept.  You can compute a table of the tradeoff in wasted bits vs maximum representable value. Or you can also realize that your computer can compute 64 bits integers... and just pick whatever precision you feel like for free pretty much.

At the end, you can do `(bits + 7) / 8`  to get the number of bytes. This last formula is simply to get the ceiling from a flooring division. And its a power of 2 division, so you can write it as `(bits + 8) >> 3`. Your compiler will optimize it for you already, and will select the most efficient solution for your CPU. So you don't really have to do it explicitly. You might be surprised to learned that some CPU can only do `>> 1`. So `>> 3` becomes three operations. It could well be that the `÷3` becomes faster at this point. Anyways, those details don't matter here.

I am not saying this is the best way to do this, but it sure is an easy way to get a decent enough precision without much effort with my limited knowledge of math :D

buchi68-0 committed:
v1.0.1

(small changes + has issues)

buchi68-0 committed:
v1.1

implemented karatsuba algorithm for multipling numbers
changed (some) functions so it's more readable by adding child functions
added negative number
added subtraction

bombela commented:
Technically the CPU returns the result into two 64bits registers.
https://www.felixcloutier.com/x86/mul

I don't think C++ gives you easily access to it besides inline assembly
(which is somewhat specific per compiler). Or using vendor specific
builtins or intrinsic
https://stackoverflow.com/questions/28868367/getting-the-high-part-of-64-bit-integer-multiplication

For what is worth, Rust makes this easier. It has a u/i128 type. And also
exposes the CPU overflow status
https://doc.rust-lang.org/std/primitive.u64.html#method.overflowing_mul and
they are exploring adding widening mul for you very use case
https://doc.rust-lang.org/std/primitive.u64.html#method.widening_mul

On Mon, Apr 28, 2025, 21:59 buchi68-0 ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In ConsoleApplication2 - Test.cpp
> <https://github.com/buchi68-0/big_numbers_in_C/pull/1#discussion_r2065454813>
> :
>
> > +#define CConsole true
> +
> +void CreateConsoleWindow() {
> +    AllocConsole();  // Create a console window
> +    FILE* fileStream;
> +    freopen_s(&fileStream, "CONOUT$", "w", stdout);  // Redirect stdout to the console
> +    freopen_s(&fileStream, "CONOUT$", "w", stderr);  // Redirect stderr to the console
> +}
> +#endif
> +#ifndef CConsole
> +#define CConsole false
> +void CreateConsoleWindow() return;
> +#endif
> +
> +typedef struct Number {
> +    uint8_t* Value;
>
> the problem I had when choosing a 64 bits limb, is that there's no 128
> bits variable to store the multiplication of the 2 numbers (or maybe just i
> don't know how to do so) ; that's why i sticked to a 32bit limb on v1.1
> Sorry for not responding before
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/buchi68-0/big_numbers_in_C/pull/1#discussion_r2065454813>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUZDFMJ4POPL64V3ADWEL234BL7AVCNFSM6AAAAABZ77TQ46VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQMBRHE4DSMRXGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

buchi68-0 committed:
v1.2.0

changed parsing function to be way faster (divided time taken by almost 20times)
changed some function param to const for better reviewness

buchi68-0 committed:
v1.3

Corrected an issue with subtraction function, where one path of the function was not returning anything (even though all operations were made)
Optimized the conversion from number to base10, effectively dividing time taken by it by about 10 times
have an option to automatically change the size of the inputs (require part and part_ to be same size ; the numbers will repeat both until the size is reached)

buchi68-0 committed:
test for new optimization

part of code to (maybe) optimize parsing using divide to conquer
reminder to refresh the project with new discoveries (details in file) that are for now only on my home's PC
will be massively tested with big or less big inputs ; and check if it's worth using

buchi68-0 committed:
Delete ConsoleApplication2 - Test.cpp

buchi68-0 committed:
Update README.md

buchi68-0 committed:
v1.4

buchi68-0 committed:
v 1.5

Made the file adapted to using with Linux.
Made the code closer to C (only things left from C++ are bools, and that's about it)
fixed some huge oversights with _Subtraction_Unsigned and _Addition_Unsigned (was checking which was bigger for _Subtraction_Unsigned but wasn't checking actual sizes, so if one number was bigger in value but not in size it was causing a seg-fault.
Some oversights are still there, and i'm working on it (trying to fix what causes error free() : invalid next size (normal) )
code is less optimized, or it's just that my actual PC is worse at single-tasking

buchi68-0 committed:
mutation toward a readable version

buchi68-0 committed:
cleaned temp files

buchi68-0 committed:
Programm fully compatible in C can be found in newly added test_C. Added subtraction and addition according to coding style. optimized subtraction and addition to be more readable

buchi68-0 committed:
changed tests to allow file usage. Max size is 10MB

buchi68-0 committed:
put multiplication back

buchi68-0 committed:
putted back karatsuba. Works for small numbers and then it just breaks. Everything is freed. Now I just have to find the problem

buchi68-0 committed:
changed main. added makefile. multiplication works again with karatsuba

buchi68-0 committed:
fixed an oversight with 0s Numbers (was putting 00 instead of 0 because of the lacking \0); removed code.cpp as I have now converted everything; fixed an issue with sign where sign was uninitialized except for one case (negative number); optimized test01.c to have more flexibility (print options are 0 for nothing, 1 for everyhing except base numbers and 2 for all); tested with very little integer and even with empty files; tested with 200kB long numbers. Planning to do some strong test soon

buchi68-0 committed:
changed test for 1MB, changed mul_numbers to be 20 lines long only

buchi68-0 committed:
remove duplicates of ConsoleApplication2.cpp; changed Makefile to put rules for clang and gcc (clang is default, gcc is a rule on it's own (gccver) and debug only is epiclang

buchi68-0 committed:
changed readme.md

buchi68-0 committed:
added comments in test01.c
changed files NUMBER1.mnb and NUMBER2.mnb
(lol I just realised these are actual valid file extension, i used it for 'My NumBer')
added a constant in main_dependencies.h that handles max length of string input from files

TODO : add division

buchi68-0 committed:
put test_C at root, put all files there, changed test01 to have timespecs to see time taken, changed readme and did new tests to get times

buchi68-0 committed:
changed Value to limbs

buchi68-0 committed:
added comments into almost all functions

buchi68-0 commented:
I saw that you're saying with the multiplication. And you're absolutely right, it's not possible to get the other part of the multiplication. Although it would make everything easier. But, I think there's a __int128_t but I'll be looking at mans (cool part of Linux btw) and all to find exactly were it lies. Also, one thing to think about is that parsing also requires an uint64_t because then I can fill 32 bits at a time, instead of bits per bits. I could potentially manage this by having a uint32_t *variable that contains a uint32_t * version of the limb (which would do the same work as before while still having a uint64_t *), but I'm not sure it would work.
I changed the program as you can see.
I think I applied everything you've told me to change, except of course for this uint64_t * limb. And well, it looks prettier now !
I added comments so it's easier to understand, but I hope it's pretty much understandable without them
Could I have a review again ? If it's not too much of me to ask...

buchi68-0 commented:
Oh yeah also, i'd like it to be compatible C99, heck even C89, so nullptr is not possible for me. Also NULL is good enough, no ? It's just (void *)0, it's also what every function that returns a pointer returns in case of problem, so I think it's good.

buchi68-0 committed:
huge type in README : 'timespec_use.c depends on main_procedures.c' which does in fact not, because the file doesn't exist. It depends on main_dependencies.h

buchi68-0 committed:
filename not following the snake lowcase convention

buchi68-0 committed:
of course I forgot to change the includes and the makefile

buchi68-0 commented:
By the way, I just found out how to do multiplications in an uint128
I'll be implementing that right now; although it might not be used right now
It's using the karatsuba alogrithm btw.

buchi68-0 committed:
fix: changed how the project behaves: memory optimisations, library instead of an executable, ...

buchi68-0 committed:
feat: added a gitignore

buchi68-0 committed:
fix: moved the test file at root

buchi68-0 committed:
feat: changed the name of some functions and added new ones

buchi68-0 committed:
feat: added the conversion string->bigint in bigint_conv.c

buchi68-0 committed:
feat: renamed addition.c for operation.c, changed the name of "multiplication" to schoolbook_multiplication, "advanced_multiplication" to multiplication and changed the size of the workspace allocated for karatsuba

buchi68-0 committed:
feat: changed the README with the newest infos

buchi68-0 committed:
feat: changed the Makefile for a new debug option

buchi68-0 committed:
fix: fixed a readme's bug

buchi68-0 committed:
fix: changed the info about the workspace allocation in README.md

buchi68-0 committed:
feat: added benchmarks for string to number and number to string conversions

buchi68-0 committed:
fix: made some micro-optimisation into the conversion from number to string

buchi68-0 committed:
feat: added a new test in main.c

buchi68-0 committed:
fix: removed unused file'

buchi68-0 committed:
breakthrough: added the toom-cook algorithm in the multiplication; crushing previous times

buchi68-0 committed:
feat: changed the test file

buchi68-0 committed:
feat: added the new features in the README

buchi68-0 committed:
fix: fixed the advertised workspace length to the actual one used

buchi68-0 committed:
feat: optimized a bit the schoolbook multiplication, giving some time saves (~2-5%)

buchi68-0 committed:
fix: changed the README, because the performances were calculated with a best case scenario

buchi68-0 committed:
feat: added an even bigger test, because now I can !

buchi68-0 committed:
feat: added times for a 40MB-O3 test

buchi68-0 committed:
breakthrough: added the fast fourier transform algorithm, complexity O(nlog(n))

buchi68-0 committed:
feat: added the new functions into the header file

buchi68-0 committed:
feat: made so the multiplication function uses FFT if the number has more than 20k limbs

buchi68-0 committed:
feat: made a more huge test, that requires ~32GB of RAM (+ SWAP)

buchi68-0 committed:
feat: added the math library into the library, needed because of FFT's complex numbers requirements

buchi68-0 committed:
feat: added new benchmarks for the FFT, and added a news for it
